src/radix_connect/app_link_parser.rs

```diff
@@ -0,0 +1,154 @@
+use crate::prelude::*;
+use url::form_urlencoded;
+use url::Url;
+
+const CONNECT_URL_PARAM_SESSION_ID: &str = "sessionId";
+const CONNECT_URL_PARAM_ORIGIN: &str = "origin";
+const CONNECT_URL_PARAM_INTERACTION_ID: &str = "interactionId";
+const CONNECT_URL: &str = "https://d1rxdfxrfmemlj.cloudfront.net";
+
+pub fn parse_mobile_connect_request(
+    url: String,
```

Comment:
Better if `parse_mobile_connect_request` takes an `impl AsRef<str>`, that way you can:

```rust
pub fn parse_mobile_connect_request(
    url: impl AsRef<Str>, // 🔮
) -> Result<MobileConnectRequest> {
    let url = url.as_ref(); // 🔮
    let connect_url = Url::parse(CONNECT_URL).unwrap();
    let parsed_url = Url::parse(url).map_err(|_| { // 🔮 no need to use `.as_str()` (you should have used there btw `&` 😊)
        CommonError::RadixConnectMobileInvalidRequestUrl {
            bad_value: url.to_owned(), // 🔮 have to make it owned (&str -> String conversion)
        }
    })?;
```

This have the advantage that we can now parse *either* a `String` or ` &str`.

And down below, in your tests, instead of:

```rust
 let connect_url = CONNECT_URL.to_owned() + "/?sessionId=123&origin=radix%3A%2F%2Fapp";
```

do this:

```rust
let connect_url = concat!(CONNECT_URL, "/?sessionId=123&origin=radix%3A%2F%2Fapp");
```

and the big win is the rest of the tests:

```rust
    #[test]
    fn url_does_not_match_expected() {
        let url = String::from("https://example.com");
        let err = parse_mobile_connect_request(url.clone()).err().unwrap();
        assert_eq!(
            err,
            CommonError::RadixConnectMobileInvalidRequestUrl { bad_value: url }
        );
    }
```

this become this:

```rust
    #[test]
    fn url_does_not_match_expected() {
        let url = "https://example.com";
        let err = parse_mobile_connect_request(url).err().unwrap(); // 🔮 No need to clone!!!
        assert_eq!(
            err,
            CommonError::RadixConnectMobileInvalidRequestUrl { bad_value: url.to_owned() } //  🔮 must call `to_owned()`
        );
    }
```